src/main/scala/org/ergoplatform/network/ErgoNodeViewSynchronizer.scala

```diff
@@ -56,31 +59,124 @@ class ErgoNodeViewSynchronizer(networkControllerRef: ActorRef,
 
-        log.info(s"${toDownload.length} persistent modifiers to be downloaded")
+    historyReaderOpt match {
+      case Some(historyReader) =>
+
+        val comparison = historyReader.compare(syncInfo)
+        log.debug(s"Comparison with $remote having starting points ${idsToString(syncInfo.startingPoints)}. " +
+          s"Comparison result is $comparison.")
+
+        val status = comparison
+        statusTracker.updateStatus(remote, status)
+
+        status match {
+          case Unknown =>
+            log.warn("Peer status is still unknown")
+          case Nonsense =>
+            log.warn("Got nonsense")
+          case Younger | Fork =>
+            val ext = historyReader.continuationIds(syncInfo, size = 400)
+            if (ext.isEmpty) log.warn("Extension is empty while comparison is younger")
+            log.info(s"Sending extension of length ${ext.length}")
+            log.debug(s"Extension ids: ${idsToString(ext)}")
+            sendExtension(remote, status, ext)
+          case Older =>
+            val ids = syncInfo.lastHeaderIds.reverse
+            val headerIds = ids.takeWhile(hId => !historyReader.isInBestChain(hId))
+            if (headerIds.nonEmpty) {
+              val maxModifiers = desiredSizeOfExpectingHeaderQueue - deliveryTracker.requestedSize
+              requestDownload(
+                maxModifiers,
+                minModifiersPerBucket = 50,
+                maxModifiersPerBucket = 400
+              )(Option(getPeersForDownloadingHeaders(remote))) { howManyPerType =>
+                Map(Header.modifierTypeId -> headerIds.reverse.filter(downloadRequired(historyReader)).take(howManyPerType))
+              }
+            }
+          case Equal => // does nothing for `Equal`
+        }
+      case _ =>
+    }
+  }
 
-        toDownload.groupBy(_._1).foreach(ids => requestDownload(ids._1, ids._2.map(_._2)))
+  /**
+    * Headers should be downloaded from an Older node, it is triggered by received sync message from an older node
+    * @param callingPeer that can be used to download headers, it must be Older
+    * @return available peers to download headers from together with the state/origin of the peer
+    */
+  private def getPeersForDownloadingHeaders(callingPeer: ConnectedPeer): (PeerSyncState, Iterable[ConnectedPeer]) =
+    statusTracker.peersByStatus.get(Older).map(PeerSyncState.Older -> _).getOrElse(PeerSyncState.OlderCalling -> Iterable(callingPeer))
+
+  /**
+    * Other persistent modifiers besides headers should be downloaded from either Older or Equal node, with fallback to Unknown or Fork
+    * @return available peers to download persistent modifiers from together with the state/origin of the peer
+    */
+  private def getPeersForDownloadingBlocks: Option[(PeerSyncState, Iterable[ConnectedPeer])] = {
+    val peersByStatus = statusTracker.peersByStatus
+    Option(peersByStatus.getOrElse(Older, Iterable.empty) ++ peersByStatus.getOrElse(Equal, Iterable.empty))
```

Comment:
Here Iterable.empty can be replaced by WrappedArray.empty which allows to avoid allocations, and way much faster.